### DISCUSSION TITLE
[RFC] Import Generated Types to Actions

### DISCUSSION THREAD

**Discussion by soorajshankar** [hasura/graphql-engine#9273] [General]
## Background

This GitHub [issue](https://github.com/hasura/graphql-engine/issues/5001) has ~100 impressions and a lot of comments.

There could be types that users want to use on Action definitions from the table schema, for example if there is a table called “authors” and if the user wants to use the type “authors” in an Action then the user needs to type this manually in the Action creation form. When there are many such types, it takes some effort and manual work to fill up the form. It also requires some learning about GraphQL types.

We [suggested](https://github.com/hasura/graphql-engine/issues/5001#issuecomment-1331962787) a possible solution that will be explained later in this document, which received some positive feedback.

## Problem

Redefining generated types in Actions is time consuming especially when there are many type definitions. Right now if a user wants to use a generated type from Hasura, they need to define it manually from the console - which is time consuming.

## Solution
While there are multiple solutions to think about, with this RFC we will discuss more of a solution that helps users to import types directly from a table shape into the Action definition.

The solution mentioned here will be available as beta from v2.17 and the purpose of this RFC is to discuss more on the enhancement plans, use cases etc. to fine tune it before GA release. 

As we already have the shape of the table in the introspection response, Hasura Console can use this to generate these types so the users can simply use it on the Action definition.

The UX would be as simple as follows, 

1. User should be able to see the following option
![image](https://user-images.githubusercontent.com/8408875/206156344-945a25ad-41ef-409b-b410-c9b0bb27a5de.png)  
2. Clicking on `From Table` would open up a new modal window as shown below
![lpgtRQ8YlKCEfIYpAdj0C7LpOwoagSlXmvooqDF2P8CvkATUZWQm-aUZzb6ELMtabSklphwihDQHVv9yJRYk16CQ9pek2tp4y-kZBHy7QJrjIh9Cyr6zJiDof3BH](https://user-images.githubusercontent.com/8408875/206156469-ee99b096-f2e0-4913-a326-bdc94d3a8f45.png)
3. Users can see the types generated from each table in real time and once all the required table shapes are selected, it can be inserted to the Action type definition input.
![Vz-Sopqz7KWhljFgN5lgC2lZU7znSbhO9w_eCIHifEzq6v4oSkIY-qR4DYqyFsW2szPWzFlQplOOV8C3F7wsqhXtUXo2V2wZIdOSNvTzBhtDCHEfxZjPhNfIpsCC](https://user-images.githubusercontent.com/8408875/206156548-4878b8a6-e533-448f-8c49-acea76b044d4.png)
4. Users can then use/modify these types on the output/input to use it in the Action definition.

## Possible Enhancements
Added separate comments (provided links to it on the following list) to each of these enhancements, so that we can identify the most asked enhancements and prioritise further. 

1. When there are relationships configured and if we want Actions also to have the dependent types, it would be nice to have a checkbox near each table to specify whether the related types also need to be imported or not. ([link](https://github.com/hasura/graphql-engine/discussions/9273#discussioncomment-4331637))
2. Right now the types imported will only have GraphQL types, and if the users want to use it in Action input types, they need to manually change it from `type` -> `input`. This might be useful when users import a lot of tables at the same time. ([link](https://github.com/hasura/graphql-engine/discussions/9273#discussioncomment-4331649))

**Comment by soorajshankar** [comment]
1. When there are relationships configured and if we want Actions also to have the dependent types, it would be nice to have a checkbox near each table to specify whether the related types also need to be imported or not.

**Comment by soorajshankar** [comment]
2. Right now the types imported will only have GraphQL types, and if the users want to use it in action input types, they need to manually change it from type -> input. This might be useful when users import a lot of tables at the same time.

**Comment by dariocravero** [comment]
While I agree that having an easy way to clone types will ease the current problem a bit I don't think it really solves it. 

In an ideal scenario I'd see the type used in the table being the same as the action's one. That way we avoid: type drifting, duplication, and potential manual syncing. 

If automatic joins with relations isn't possible/too complex, could a stop gap be that Hasura leverages the same mechanism used in remote schemas to connect results back to the db at least?

**Comment by rahulagarwal13** [comment]
@dariocravero  - thanks for your comment. Would like to know the scenario when your database type changes, would you like your Action definitions to be in sync with that type change or would you expect to change them manually. And will such changes in your database schema would cause any disruption in clients who are using the dependent actions.

We would like to update you that [reusing enums in Actions](https://github.com/hasura/graphql-engine/issues/4513) is on the roadmap. Does your use-case involve enums? If not what kind of types are you looking to reuse?

**Reply by erquhart** [reply]
I think the type staying in sync would be the default expectation - a named type is assumed to have a single definition. I'm trying to use actions to do some validation that isn't possible with postgres triggers, but with a properly typed Typescript application using graphql-codegen, the difference in `__typename` alone is a serious blocker. I don't want to create any new types at all, just want to reuse existing ones.

**Reply by dariocravero** [reply]
@rahulagarwal13 I agree with @erquhart. I'd expect the types to be in sync too.

Regarding the changes causing issues to clients, when paired with DB tables, I see actions as extensions of those tables. So the effect would be the same as if we were to make changes to the tables too. It's a conscious decision to link the two. Otherwise, I'd just use a different type altogether for the action.

Reusing enums would also be useful!
